### PR TITLE
Batch 1: Github Actions, Brew recipe, linting suggestions, fish completion etc

### LINKED ISSUE #3: Contributing back to tome
👋 

Long time `sub` fan and found your project thanks to a friend. Thanks for setting up a new incarnation! I've setup a `sub` based cli at the last couple places I've worked and it's been great. Interested to play with tome and see if it's my new goto.

When trying to run the current HEAD of tome, I get a test failure:

```
bash-5.1$ cargo test
    Finished test [unoptimized + debuginfo] target(s) in 0.01s
     Running target/debug/deps/tome-2fa1d8bb501aff4b

running 16 tests
test commands::tests::test_should_not_source ... ok
test commands::tests::test_usage ... ok
test commands::tests::test_should_source ... ok
test commands::tests::test_help ... ok
test tests::test_dangerous_characters_quoted ... ok
test tests::test_script_directory_argument ... ok
test tests::test_directory_completion ... ok
test tests::test_script_in_directory ... ok
test tests::test_script_in_directory_not_found ... ok
test tests::test_root_directory_completion ... ok
test tests::test_simple_script ... ok
test tests::test_source ... ok
test tests::test_use_arg ... ok
test tests::test_help_page ... ok
test tests::test_simple_script_completion ... ok
test tests::test_source_completion ... FAILED

failures:

---- tests::test_source_completion stdout ----
thread 'tests::test_source_completion' panicked at 'assertion failed: `(left == right)`
  left: `Ok("")`,
 right: `Ok("foo baz\n")`', src/tests.rs:43:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    tests::test_source_completion

test result: FAILED. 15 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.17s
```

I investigated this ~by setting up a script to recompile tome and prepend the compiled ./target/debug/tome` to my path (otherwise system tome would be used) and~ (edited due to my misunderstanding test harness) re-run the tests across the various git tags. In my testing this one test has failed from v0.1 onwards to v0.7.1.

A couple questions:
* Are you seeing different test results than I am?
* Is the desired behavior still that `source` style scripts allow for completion?
* If I'm hacking on this, would you appreciate things like Github Actions for CI and patches for that test failure?

### PR DESCRIPTION
**Pull request by zph**
Closes #3.

See #3 for starting context. Then I left inline github comments and resolved the ones that are `informational only` and left unresolved the ones that need and answer or change.

Changes:
- Adds brew formula
- Addresses linter complaints (`cargo clippy`)
- Adds Github Actions workflow for `cargo test` and `cargo clippy`. But requires repo config to enable in Github UI.
- Adds docs and autocomplete for fish shell 🤩 
- Fixes panic during unwrap()
- Completes naming update from cookbook->tome
- Symlinks docs/index.md to README.md to avoid duplication and improve docs

TODO pre-merge:
- [x] Address todos I left in code or Github comments, ie unresolved comments.
- [x] Address TODO in Homebrew Recipe (https://github.com/toumorokoshi/tome/pull/4/files#diff-93a97375ec7323face1d94029d385da54c155f4516a25d567b630654efe211f1R19)
- [x] Remove repetition https://github.com/toumorokoshi/tome/pull/4/files#r598067902
- [x] Add License file and update brew recipe with it. (Added proposed MIT, can revise if diff one is preferred)
- [x] Decide on .cargo/bin and how to use ENV vars for building out of the box on both darwin and linux
- [x] Manual regression testing
- [x] Delete ignored test as invalid?

TODO post-merge:
- [ ] Enable Github actions

### CODE REVIEW

File: .cargo/config [hunk 1]
```diff
@@ -1,3 +1,7 @@
-[target.x86_64-apple-darwin]
-linker = "x86_64-apple-darwin15-clang"
-ar = "x86_64-apple-darwin15-ar"
\ No newline at end of file
+# HACK: Disabled to allow for building on Darwin
```
Comment (zph) [inline .cargo/config:1]: Unsure how we want to handle this.

Change was because the current setup breaks compilation for me on Darwin for Darwin target. I can override using ENV vars and get it working or comment out 🤔.

My recommendation is to provide people with the right ENV vars they can set depending on their build platform to `DO THE RIGHT THING`. Then we remove this file.

Sound good?

Comment (zph) [reply .cargo/config:1]: I could cherry pick those changes over from Batch #2.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -0,0 +1,22 @@
+on: [push]
```
Comment (zph) [inline .github/workflows/build.yml:1]: TODO: works but needs repo configuration in Github.

File: .vscode/launch.json [hunk 3]
```diff
@@ -7,12 +7,12 @@
         {
             "type": "lldb",
             "request": "launch",
-            "name": "Debug executable 'cookbook'",
```
Comment (zph) [inline .vscode/launch.json:10]: Renames

File: HomebrewFormula [hunk 4]
```diff
@@ -0,0 +1 @@
+pkg/brew
```
Comment (zph) [inline HomebrewFormula:1]: Symlink to allow for using custom folder as the brew formula location. Hat tip to ... ripgrep for this pattern?

File: docs/index.md [hunk 5]
```diff
@@ -84,6 +84,12 @@ to be named "my-commands" you'd put the following in your .bashrc or .zshrc:
 eval "`~/bin/tome init my-commands ~/my-scripts $0`"
 ```
 
+For fish shell:
+```
+# in ~/.config/fish/conf.d/tome.fish
+tome init sc ~/my_script_dir $0 | source
```
Comment (zph) [inline docs/index.md:90]: Fish requires different sourcing command.

File: pkg/brew/tome.rb [hunk 6]
```diff
@@ -0,0 +1,25 @@
+class Tome < Formula
+  HOMEPAGE = "https://github.com/toumorokoshi/tome"
+
+  desc "Modern replacement for 'sub'"
+  homepage HOMEPAGE
+  url HOMEPAGE
+  # TODO: What's correct license?
```
Comment (zph) [inline pkg/brew/tome.rb:7]: @toumorokoshi Want me to add a license file to project, if so, what? I have a preference towards MIT or BSD if those are ones you approve.

File: pkg/brew/tome.rb [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+  head "#{HOMEPAGE}.git"
+
+  depends_on "rust" => :build
+
+  def install
+    system "cargo", "install", *std_cargo_args
+  end
+
+  test do
+    # TODO: setup testing
```
Comment (zph) [inline pkg/brew/tome.rb:21]: TODO @zph

File: src/commands/help.rs [hunk 8]
```diff
@@ -21,11 +21,11 @@ pub fn help(root: &str, mut _args: Peekable<Iter<String>>) -> io::Result<String>
             escape_slashes(&script.summary_string)
         ))
     }
-    return Ok(format!(
```
Comment (zph) [inline src/commands/help.rs:24]: Linting warning fix

File: src/commands/init.rs [hunk 9]
```diff
@@ -1,6 +1,8 @@
 use std::{
     iter::{Iterator, Peekable},
     slice::Iter,
+    env,
+
```
Comment (zph) [inline src/commands/init.rs:5]: TODO format to remove spare line
```suggestion
```

File: src/commands/init.rs [hunk 10]
```diff
@@ -10,9 +12,13 @@ macro_rules! init_help_body {
...
 
-An example cookbook init invocation looks like:
+An example tome init invocation looks like:
+
+# Bash/Zsh
+source <(tome init sc ~/my_script_dir $0)
 
-source <(./cookbook init sc ~/my_script_dir $0)
+# Fish
+tome init sc ~/my_script_dir $0 | source
```
Comment (zph) [inline src/commands/init.rs:21]: Meta: I'd like one place we define these and then re-use in various locations. As it stands now I recall 2 places these exist. One in `.rs` and one in `.md`

File: src/commands/init.rs [hunk 11]
```diff
@@ -77,10 +83,88 @@ complete -F _{function_name}_completions {function_name}
...
 }
 
-// given the location of the cookbook executable, return
-// back the init script for cookbook.
 
-pub fn init(cookbook_executable: &str, mut args: Peekable<Iter<String>>) -> Result<String, String> {
+macro_rules! fish_init_body {
+    () => {
+        r#"
+function __fish_tome_help_message
```
Comment (zph) [inline src/commands/init.rs:90]: Fish shell is really nice and simple for completions. They're standard functions and they are pretty easy to debug. Also the UI of it is pretty ⭐

File: src/commands/init.rs [hunk 12]
```diff
@@ -77,10 +83,88 @@ complete -F _{function_name}_completions {function_name}
...
+end
+complete -c {function_name} -f -a "(__fish_tome_completion_fn {script_root} $argv)"
+# End tome alias
+"#
+    };
+}
+
+// given the location of the tome executable, return
+// back the init script for tome.
+
```
Comment (zph) [inline src/commands/init.rs:166]:
```suggestion
```

File: src/commands/init.rs [hunk 13]
```diff
@@ -99,28 +183,49 @@ pub fn init(cookbook_executable: &str, mut args: Peekable<Iter<String>>) -> Resu
             ))
         }
     };
-    let _shell_type = match args.next() {
+    let shell_env = match env::var("SHELL") {
```
Comment (zph) [inline src/commands/init.rs:186]: We can't rely on `$0` because it's undefined in fish shell, so fallback to `$SHELL`

Comment (zph) [reply src/commands/init.rs:186]: Alternatively, I like requiring people in the invocation to use ``` tome init sc ./example fish | source```. That's entirely reasonable and has precedent in direnv, for example.

File: src/commands/init.rs [hunk 14]
```diff
@@ -99,28 +183,49 @@ pub fn init(cookbook_executable: &str, mut args: Peekable<Iter<String>>) -> Resu
...
     // current environment (such as cd you into a specific)
     // directory.
-    // TODO: add conditionals if other shells need different support
-    Ok(format!(
-        bash_zsh_init_body!(),
-        cookbook_executable = cookbook_executable,
-        script_root = script_root,
-        function_name = function_name
-    ))
+    match shell_type {
```
Comment (zph) [inline src/commands/init.rs:212]: This is verbose, but I couldn't figure out a more dynamic way to do so when the init_body!() executions are macros 🤔

File: src/commands/script.rs [hunk 15]
```diff
@@ -49,25 +49,23 @@ impl Script {
             if consuming_help {
                 if line.starts_with("# END HELP") {
                     consuming_help = false;
-                } else if line.starts_with("# ") {
+                } else if let Some(rest) = line.strip_prefix("# ") {
```
Comment (zph) [inline src/commands/script.rs:52]: linting suggestion

File: src/commands/script.rs [hunk 16]
```diff
@@ -139,7 +137,7 @@ impl Script {
                     arg.push('\'');
                     escaped_command_string.push(arg);
                 }
-                Ok(escaped_command_string.join(" ").to_owned())
```
Comment (zph) [inline src/commands/script.rs:142]: linting suggestion from `cargo clippy`

File: src/directory.rs [hunk 17]
```diff
@@ -47,10 +47,10 @@ pub fn is_tome_script_directory(dir: &Path) -> bool {
...
         .to_str()
         .unwrap_or_default()
-        .starts_with(".")
+        .starts_with('.')
     {
         return false;
     }
     tomeignore_location.push(".tomeignore");
-    return !tomeignore_location.exists();
+    !tomeignore_location.exists()
```
Comment (zph) [inline src/directory.rs:55]: cargo clippy

File: src/tests.rs [hunk 18]
```diff
@@ -39,6 +39,7 @@ fn test_source() {
 }
 
 #[test]
+#[ignore]
```
Comment (zph) [inline src/tests.rs:42]: Ignored because during a git bisect I couldn't find a time this succeeded anywhere in git history.

@toumorokoshi What's desired behavior of this? Can you think of sourced files that we need `--completion` for?

Comment (zph) [reply src/tests.rs:42]: My recommendation is code is behaving as desired and test describes undesired behavior.

**Review by zph — COMMENTED** [review]
Inline comments on everything. I'll update the PR description and then I need to pull this specifically down and do manual regression testing.

File: .cargo/config (hunk 1 above)
Comment (zph) [reply .cargo/config:1]: Perhaps this is best solved by either `darwin.env` and `linux.env` files you can source and then your shell is setup to work right. Or, a shell script in `bin/setenv` that outputs the right env depending on which build system you run :D.

TODO: then update docs to indicate dev environment setup.

**Comment by zph** [comment]
In absence of CI: 
image

**Comment by zph** [comment]
@toumorokoshi Ready for initial review while I work on finishing the TODOs 🥇

File: .cargo/config (hunk 1 above)
Comment (zph) [reply .cargo/config]: See ./bin/release and ./bin/setenv and docs/index.md under "Cross Compilation" for info 👯 

DONE

File: src/tests.rs (hunk 18 above)
Comment (zph) [reply src/tests.rs]: (deleted but can revisit)

**Comment by zph** [comment]
Ok, in a clean shell, using local build, it's behaving well for fish shell & tests are still passing.

File: LICENSE.md [hunk 19]
```diff
@@ -0,0 +1,21 @@
+MIT License
+
+Copyright (c) 2021 Yusuke Tsutsumi
```
Comment (toumorokoshi) [inline LICENSE.md:3]: should this be 2020? I guess it doesn't matter too much.

Also thank you! was going to add a license, but this LGTM.

File: docs/index.md [hunk 20]
```diff
@@ -81,12 +91,49 @@ to be named "my-commands" you'd put the following in your .bashrc or .zshrc:
 
 ```
 # posix example, e.g. .bashrc
-eval "`~/bin/tome init my-commands ~/my-scripts $0`"
+eval "$(tome init my-commands ~/my-scripts bash)"
+
+# zsh
```
Comment (toumorokoshi) [inline docs/index.md:96]: nitpick: should the "zsh" and "bash" be in their own code blocks as well? I see fish is the odd one out.

File: docs/index.md [hunk 21]
```diff
@@ -81,12 +91,49 @@ to be named "my-commands" you'd put the following in your .bashrc or .zshrc:
...
 *NOTE*: make sure to include the double quotes with the nested backticks. This ensures that newlines are captured
 and evaluated appropriately.
 
 ### 4. Start a New Shell
 
-Once installed, start a new shell and you should have your new command!
\ No newline at end of file
+Once installed, start a new shell and you should have your new command!
+
+## Developing Tome
```
Comment (toumorokoshi) [inline docs/index.md:113]: might consider moving this to a `CONTRIBUTING.md`.

File: pkg/brew/tome.rb [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+  desc "Modern replacement for 'sub'"
+  homepage HOMEPAGE
+  url HOMEPAGE
+  license "MIT"
+  version '0.1.0'
+  revision 1
+
+  head "#{HOMEPAGE}.git"
+
+  depends_on "rust" => :build
```
Comment (toumorokoshi) [inline pkg/brew/tome.rb:13]: It might be a heavy dependency for some users to require rust to install the relatively small tome.

Is it acceptable to download a release instead?

File: pkg/brew/tome.rb [hunk 23]
```diff
@@ -0,0 +1,24 @@
+class Tome < Formula
+  HOMEPAGE = "https://github.com/toumorokoshi/tome"
+
+  desc "Modern replacement for 'sub'"
+  homepage HOMEPAGE
+  url HOMEPAGE
+  license "MIT"
+  version '0.1.0'
```
Comment (toumorokoshi) [inline pkg/brew/tome.rb:8]: in these situations, should the version match the version of the binary released on GitHub?

File: src/commands/init.rs [hunk 24]
```diff
@@ -77,10 +82,87 @@ complete -F _{function_name}_completions {function_name}
     };
 }
 
-// given the location of the cookbook executable, return
-// back the init script for cookbook.
+macro_rules! fish_init_body {
```
Comment (toumorokoshi) [inline src/commands/init.rs:85]: It might be nice to replace these macro bodies with something like [include_str](https://doc.rust-lang.org/std/macro.include_str.html).

File: src/commands/init.rs (hunk 13 above)
Comment (toumorokoshi) [reply src/commands/init.rs:184]: yeah, I think it's more robust to have the invoker pass the shell, rather than implicitly assume some environment variable.

That's how it was originally, correct? you just pass in the shell type during init.

File: src/commands/init.rs (hunk 14 above)
Comment (toumorokoshi) [reply src/commands/init.rs:210]: you could create a block just to extract the template, then pass that into the format argument. Something like:

```
let shell_template = match shell_type {
    "fish" => fish_init_body!(),
    ....,
    _ => {return Err()}
}

return Ok(format!(shell_template, tome_executable=tome_executable,...))  
```

File: src/commands/script.rs [hunk 25]
```diff
@@ -49,25 +49,23 @@ impl Script {
...
-                    // as it can indicate the command to run the script with.
-                    // metadata lines must be consecutive.
-                    break;
+                    help_string.push_str(rest);
                 }
+            } else if line.starts_with("# SOURCE") {
+                should_source = true;
+            } else if line.starts_with("# START HELP") {
+                consuming_help = true;
+            } else if line.starts_with("# SUMMARY: ") {
```
Comment (toumorokoshi) [inline src/commands/script.rs:61]: nitpick: should this also use strip_prefix? IIUC the above strip_prefix was introduced to eliminate the hand-counting of a starting index for push_str.

File: src/directory.rs (hunk 17 above)
Comment (toumorokoshi) [reply src/directory.rs:55]: noisy fellow, isn't clippy? thanks for fixing these up!

File: src/tests.rs [hunk 26]
```diff
@@ -38,19 +38,6 @@ fn test_source() {
     );
 }
 
-#[test]
-fn test_source_completion() {
```
Comment (toumorokoshi) [inline src/tests.rs:42]: why was this test removed?

**Review by toumorokoshi — CHANGES_REQUESTED** [review]
Overall looks good! a couple minor comments, but the main ones are:

- switching back the shell detector behavior to being passed in, vs using the $SHELL variable.
- question regarding a deleted test, when functionality looks to be the same.

Thank you for all the improvements!

File: src/tests.rs (hunk 18 above)
Comment (toumorokoshi) [reply src/tests.rs]: source completion is important, since even if the final execution ends up being sourced, you still want to be able to provide auto-completion for things like arguments to an eventually sourced function.

one example is [a shortcut to a directory where you want to allow tab-completion of sub-directories and files](https://github.com/toumorokoshi/ytlaces/blob/master/files/cookbook/w).

Comment (toumorokoshi) [reply src/tests.rs]: but I heard the test has been failing for all of history though. I think it's fine to merge this in deleted to fix the build, then pick up a follow-up PR (which I can own) to merge it back in passing.

File: LICENSE.md (hunk 19 above)
Comment (zph) [reply LICENSE.md:3]: :D. Now that I think about it, it might be more accurately as 2019-2021 YOU.
```suggestion
Copyright (c) 2019-2021 Yusuke Tsutsumi
```

File: docs/index.md (hunk 20 above)
Comment (zph) [reply docs/index.md:96]: Yes! I think I did that in the later PR.

Ok w/ leaving it as a nit that we address later in a docs cleanup task? If so, I'll add a placeholder issue to make a pass at documentation cleanup when most interface changes have landed.

File: docs/index.md (hunk 21 above)
Comment (zph) [reply docs/index.md:113]: Good idea, will do

File: pkg/brew/tome.rb (hunk 22 above)
Comment (zph) [reply pkg/brew/tome.rb:13]: We can do that also. People want/expect the ability to compile if desired.

What we could do is setup homebrew to either get most recent release, OR compile from scratch if wanting it built from master.

If you like that approach, I'll break out a follow up task to address.

File: pkg/brew/tome.rb (hunk 23 above)
Comment (zph) [reply pkg/brew/tome.rb:8]: Yes. Which will require keeping version synchronized in two place.

I could break out a task to make a `version_bump` helper script that updates the version in both places 🤔 to make sure it reliably happens?

File: src/commands/init.rs (hunk 24 above)
Comment (zph) [reply src/commands/init.rs:85]: Ok, want that tweak in this PR or follow on issue?

**Comment by zph** [comment]
re: SHELL variable...

I'd like your input. The constraint is: when initializing the `tome init cmd ./example $0` that $0 only works in zsh and bash, not in fish. I built in the `look at ENV var SHELL` which is default set in terminal as a fallback so zsh, bash and fish could all try to use the same interface.

But... my preference is to require someone to explicitly declare their shell, like you're saying. This removes need for `$SHELL` fallback and makes the initialization command consistent over all shells, ie:

```
tome init cmd ./example [bash|zsh|fish]
```

There's precedent for requiring the manual declaration in tools like direnv's init interface :).

TLDR: I'll happily make it and docs match with having people declare what shell.

Does that work well for you?

File: src/commands/script.rs (hunk 25 above)
Comment (zph) [reply src/commands/script.rs:61]: Good nit :). But if you're ok w/ it I think I already made changes to this in Batch #2 and prefer that we nit it over there 🗡️

File: src/commands/init.rs (hunk 13 above)
Comment (zph) [reply src/commands/init.rs:184]: See here: https://github.com/toumorokoshi/tome/pull/4/files#diff-cadd09612a7d662d0ed12de2a0c70dc7afbea48e3aa5861398fea744dc49f476R188

After my change, the logical flow is:
1. Try the existing lookup of `$0` which is the args.next() on L188.
2. If Some, use it
3. If None, then fallback to using `$SHELL` env var which is automatically set in terminal :).

I didn't drop your existing way :D. But I did add a fallback for when that would result in an error with fish.

But but but... see my longer comment on PR for my overall suggestion that hews closer to what you're asking for and adds commandline interface consistency.

File: src/commands/init.rs (hunk 14 above)
Comment (zph) [reply src/commands/init.rs:210]: Of course :D I'll try that.

File: .github/workflows/build.yml (hunk 2 above)
Comment (zph) [reply .github/workflows/build.yml:1]: Promoted to separate issue.

File: src/commands/init.rs (hunk 13 above)
Comment (zph) [reply src/commands/init.rs:184]: (Weird, this was folded strangely, so I responded directly to your comment in a void 🥼)

File: docs/index.md (hunk 21 above)
Comment (zph) [reply docs/index.md:113]: Done!

File: pkg/brew/tome.rb (hunk 22 above)
Comment (zph) [reply pkg/brew/tome.rb:13]: Broke into new issue.

File: pkg/brew/tome.rb (hunk 23 above)
Comment (zph) [reply pkg/brew/tome.rb:8]: Promoted to new issue.

**Comment by toumorokoshi** [comment]
> There's precedent for requiring the manual declaration in tools like direnv's init interface :).
>TLDR: I'll happily make it and docs match with having people declare what shell.
> Does that work well for you?

Works for me! I don't think `SHELL` is reliable, as that just dictates the user's login shell, rather than the current working shell.

See the following from `man login` on Linux:

```
    The user and group ID will be set according to their values in the /etc/passwd file.  There is one exception if the user ID is zero: in this case, only the primary group ID of the account is set.  This should al‐
       low the system administrator to login even in case of network problems.  The value for $HOME, $USER, $SHELL, $PATH, $LOGNAME, and $MAIL are set according to the appropriate fields in the  password  entry.   $PATH
       defaults to /usr/local/bin:/bin:/usr/bin for normal users, and to /usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin for root, if not otherwise configured.
```

Add an example of `SHELL` misreporting on my machine:

```
tsutsumi:/home/tsutsumi (0)
$ /bin/bash
[tsutsumi@tsutsumi-laptop ~]$ echo $SHELL
/bin/zsh
```

File: src/commands/init.rs (hunk 24 above)
Comment (toumorokoshi) [reply src/commands/init.rs:85]: follow on is fine, non-blocking comment.

Comment (toumorokoshi) [reply src/commands/init.rs:85]: https://github.com/toumorokoshi/tome/issues/11

File: src/commands/init.rs (hunk 13 above)
Comment (toumorokoshi) [reply src/commands/init.rs:184]: ah ok, I see what you did. It sounds like you're planning on switching to always passing in a shell argument which is good. If you want to merge this in and address that in a future PR, that's fine too.

File: src/commands/init.rs (hunk 14 above)
Comment (zph) [reply src/commands/init.rs:210]: I tried that approach and was having type issues 🤔. 

I can try again and then update with my findings.

File: src/tests.rs (hunk 18 above)
Comment (zph) [reply src/tests.rs]: Addressed ⚡

File: bin/release [hunk 27]
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env bash
+
+set -eou pipefail
+set -x
```
Comment (zph) [inline bin/release:4]:
```suggestion
```

File: src/commands/init.rs (hunk 13 above)
Comment (zph) [reply src/commands/init.rs:184]: 👍 ok, I'll do this in the future PR.

**Comment by zph** [comment]
@toumorokoshi `Ok("yay")`.

The tests are passing, the TODOs from this PR are either
- addressed
- promoted out to their own issues
- part of batch 2

Except removing the duplication in init.rs among the shells. I get stuck on type issues still. The closest I came was in 588bbd6 but then the `{function_name}` interpolation wasn't happening. I welcome your patch but I've beaten my novice rust chops against it unsuccessfully.

Aside from that:

:D I'm ready for you to take a final look through this and then we can merge. I think I'll have time in next 3 days to work on Batch #2 which is what I'm running for myself right now.

Can you please pull this down and try it out for yourself?

File: src/commands/init.rs (hunk 14 above)
Comment (toumorokoshi) [reply src/commands/init.rs:210]: I wouldn't worry too much about it, it's a minor optimization that doesn't improve readability by that much, and correctness at all :) 

Thanks for trying!

**Review by toumorokoshi — APPROVED** [review]
LGTM! let's merge this in.
